### PR TITLE
mgr/dashboard: Enable compiler options used by Angular --strict flag

### PR DESCRIPTION
**Pull request by tspmelo**
Recently Angular added a new flag to the "ng new" command, that adds 5 new typescript compiler options.
They add more strict rules that help the code to be more readable and less prone to errors.
The PR can be found here: https://github.com/angular/angular-cli/pull/14905

Rules that will be applied in this PR:
**noImplicitAny** - https://basarat.gitbooks.io/typescript/docs/options/noImplicitAny.html
**noImplicitReturns** - https://lucidchart.com/techblog/2018/06/20/how-to-actually-improve-type-safety-with-the-typescript-strict-flags/
**noImplicitThis** - https://logicbig.com/tutorials/misc/typescript/no-implicit-this.html
**noFallthroughCasesInSwitch** - https://typescriptlang.org/docs/handbook/release-notes/typescript-1-8.html

**strictNullChecks** will be applied in a later PR.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tspmelo** [comment]
jenkins test dashboard

**Comment by tspmelo** [comment]
jenkins test make check

**Comment by epuertat** [comment]
This is a real admirable work @tspmelo ! Did you use some script to automate it?

On the use of `CdObj<T>`, while it at least types the variable as a mapping, it appears >190 times as `CdObj<any>`, hence not performing any type check at all (beyond the mapping interface). I was wondering whether the increase in type safety is worthy the amount of changes required (as this massive clean-ups make backports harder to perform).

An alternative would be to use a Typescript utility type, like `Record<K, T>`. It would work as an alias for `CdObj<any>` with `Record<string, any>`. This way we could also support fast incremental typing: `Record<"field1" | "field2" | "field3", string>`?

**Comment by tspmelo** [comment]
Thanks, most of it was manually or some search and replace.

The only reason I created the `CdObj` was to be able to use the mapping interface, since with the stricter options you can't used it anymore with `object`s.
I added the Generic to be able to be more strict in some situations, but it ended up being just few cases.

I tried other utilities types but none worked with using string as an index.
Didn't checked `Record` and that one seems to work :)

I just did a new search on ways to fix this problem and found that there is a compiler option that might help us, although reducing the strictness of the compiler.

I see this possible solutions:

1. Enable `suppressImplicitAnyIndexErrors `option.
This will allows us to access objects with a string index.
This will be like before and I could remove the use of `CdObj<T>` and remove all the `[k: string]: any;` I added to classes/interfaces.
This would reduce a lot of the changes made.
2. Replace all use of `CdObj<T>` with `Record<K, T>`
This will remove the need to import a new file while keeping all functionallity.
Still needs `[k: string]: any;` to be added to classes/interfaces.
3. Apply both 1 and 2.
We would apply option 1, but on situations where we have `CdObj<T>` that uses a non `any` type, we could use `Record<K, T>` to keep the strictness.

My vote is on the 3rd option.

**Comment by epuertat** [comment]
@tspmelo, I'd take the option that:
1. Increases quality of existing code,
2. Enforces new code to meet the stricter rule,
3. Minimizes changes (for maintained back-portability).

I think your suggestion #3 fulfills all of those, so I'm happy with that.

**Comment by tspmelo** [comment]
I have removed the use of `CdObj` and all new `[k: string]: any;`

**Comment by tspmelo** [comment]
jenkins test dashboard

**Comment by callithea** [comment]
jenkins test make check

**Comment by tspmelo** [comment]
jenkins test dashboard

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/crushmap/crushmap.component.ts [hunk 1]
```diff
@@ -74,7 +75,8 @@ export class CrushmapComponent implements OnInit {
   }
 
   onNodeSelected(e: NodeEvent) {
-    const { name, type, status, ...remain } = this.metadataKeyMap[e.node.id];
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/crushmap/crushmap.component.ts]: Could we use `as number` instead of importing `lodash` just to make the compiler happy?
If yes I would prefer `as number` instead.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.spec.ts [hunk 2]
```diff
@@ -59,11 +60,11 @@ describe('RbdFormComponent', () => {
...
-    let createAction;
-    let editAction;
-    let cloneAction;
-    let copyAction;
-    let rbdServiceGetSpy;
+    let createAction: any;
+    let editAction: any;
+    let cloneAction: any;
+    let copyAction: any;
+    let rbdServiceGetSpy: any;
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.spec.ts]:
```suggestion
    let rbdServiceGetSpy: jasmine.Spy;
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-directories/cephfs-directories.component.spec.ts [hunk 3]
```diff
@@ -35,15 +35,15 @@ describe('CephfsDirectoriesComponent', () => {
...
-  let lsDirSpy;
-  let modalShowSpy;
-  let notificationShowSpy;
-  let minValidator;
-  let maxValidator;
-  let minBinaryValidator;
-  let maxBinaryValidator;
-  let originalDate;
-  let modal;
+  let lsDirSpy: any;
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-directories/cephfs-directories.component.spec.ts]:
```suggestion
  let lsDirSpy: jasmine.Spy;
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.spec.ts [hunk 4]
```diff
@@ -20,7 +20,7 @@ import { HostsComponent } from './hosts.component';
 describe('HostsComponent', () => {
   let component: HostsComponent;
   let fixture: ComponentFixture<HostsComponent>;
-  let hostListSpy;
+  let hostListSpy: any;
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.spec.ts]:
```suggestion
  let hostListSpy: jasmine.Spy;
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-details/osd-details.component.spec.ts [hunk 5]
```diff
@@ -20,7 +20,7 @@ describe('OsdDetailsComponent', () => {
   let fixture: ComponentFixture<OsdDetailsComponent>;
   let debugElement: DebugElement;
   let osdService: OsdService;
-  let getDetailsSpy;
+  let getDetailsSpy: any;
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-details/osd-details.component.spec.ts]:
```suggestion
  let getDetailsSpy: jasmine.Spy;
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/prometheus/silence-form/silence-form.component.spec.ts [hunk 6]
```diff
@@ -41,15 +41,15 @@ describe('SilenceFormComponent', () => {
   let notificationService: NotificationService;
   let router: Router;
   // Spies
-  let rulesSpy;
-  let ifPrometheusSpy;
+  let rulesSpy: any;
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/prometheus/silence-form/silence-form.component.spec.ts]:
```suggestion
  let rulesSpy: jasmine.Spy;
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/prometheus/silence-form/silence-form.component.spec.ts [hunk 7]
```diff
@@ -174,7 +174,7 @@ describe('SilenceFormComponent', () => {
 
   describe('redirect not allowed users', () => {
     let prometheusPermissions: Permission;
-    let navigateSpy;
+    let navigateSpy: any;
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/prometheus/silence-form/silence-form.component.spec.ts]:
```suggestion
    let navigateSpy: jasmine.Spy;
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/notification.service.ts [hunk 8]
```diff
@@ -171,7 +171,24 @@ export class NotificationService {
...
+        toastrCall = this.toastr.error;
+        break;
+      case NotificationType.info:
+        toastrCall = this.toastr.info;
+        break;
+      case NotificationType.success:
+        toastrCall = this.toastr.success;
+        break;
+      default:
+        return;
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/notification.service.ts]: This doesn't seem code style related, it also makes the code less readable and produces a state that is never be used and if it were used (which should raise an error) no error would be thrown.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert-formatter.ts [hunk 9]
```diff
@@ -65,7 +65,8 @@ export class PrometheusAlertFormatter {
       info: ['suppressed', 'unprocessed'],
       success: ['resolved']
     };
-    return NotificationType[_.findKey(types, (type) => type.includes(status))];
+    const key = _.findKey(types, (type) => type.includes(status));
+    return _.get(NotificationType, key);
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert-formatter.ts]: I wonder why this change is needed, as I don't see the difference and it's not more type save than before. The only difference is that `_.get` is taking a view more steps.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-configuration-form/rbd-configuration-form.component.ts [hunk 10]
```diff
@@ -77,10 +81,15 @@ export class RbdConfigurationFormComponent implements OnInit {
     const result = {};
 
     this.rbdConfigurationService.getWritableOptionFields().forEach((config) => {
-      const control = this.form.get('configuration').get(config.name);
+      const control: any = this.form.get('configuration').get(config.name);
       const dirty = control.dirty;
 
-      if (this.initialData && this.initialData[config.name] === control.value) {
```
[the remark this answers is not in the document]
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-configuration-form/rbd-configuration-form.component.ts]: Hi, I think this caused an error before I enabled `suppressImplicitAnyIndexErrors`.
I will revert it since its no longer the case.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/crushmap/crushmap.component.ts (hunk 1 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/crushmap/crushmap.component.ts]: Same as before, will revert it.

File: src/pybind/mgr/dashboard/frontend/e2e/ui/notification.e2e-spec.ts [hunk 11]
```diff
@@ -20,7 +20,7 @@ describe('Notification page', () => {
   });
 
   it('should open notification sidebar', async () => {
-    await notification.waitStaleness(notification.getSidebar());
+    await notification.waitInvisibility(notification.getSidebar());
```
[the remark this answers is not in the document]
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/e2e/ui/notification.e2e-spec.ts]: This was just to test the fix for the e2e, already created a separate PR, will remove it.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.spec.ts (hunk 2 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.spec.ts]: Done

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/pwd-expiration-notification/pwd-expiration-notification.component.ts [hunk 12]
```diff
@@ -37,11 +37,9 @@ export class PwdExpirationNotificationComponent implements OnInit {
...
 
   private getExpirationDays(pwdExpirationDate: number): number {
-    if (pwdExpirationDate) {
-      const current = new Date();
-      const expiration = new Date(pwdExpirationDate * 1000);
-      return Math.floor((expiration.valueOf() - current.valueOf()) / (1000 * 3600 * 24));
-    }
+    const current = new Date();
+    const expiration = new Date(pwdExpirationDate * 1000);
+    return Math.floor((expiration.valueOf() - current.valueOf()) / (1000 * 3600 * 24));
```
[the remark this answers is not in the document]
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/components/pwd-expiration-notification/pwd-expiration-notification.component.ts]: If I revert this I will need to add `return undefined;` at the end of the function.
I though this was a better fix.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts [hunk 13]
```diff
@@ -304,7 +304,7 @@ export class TableComponent implements AfterContentChecked, OnInit, OnChanges, O
     };
     return columns
       .reduce(
-        (result, value, index) =>
+        (result, value, index: number) =>
```
[the remark this answers is not in the document]
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts]: Looks like it does.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert-formatter.ts (hunk 9 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert-formatter.ts]: With `noImplicitAny `enabled and `suppressImplicitAnyIndexErrors` disabled, TS is very strict with the type of the index used.
p.e. in a normal class you can't use any index to access its properties.

Since this is no longer the case I will revert.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts [hunk 14]
```diff
@@ -411,20 +411,24 @@ export class PoolFormComponent implements OnInit {
...
 
-  private replicatedPgCalc(pgs): number {
+  private replicatedPgCalc(pgs: number): number {
     const sizeControl = this.form.get('size');
     const size = sizeControl.value;
     if (sizeControl.valid && size > 0) {
       return pgs / size;
     }
+
+    return undefined;
```
[the remark this answers is not in the document]
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: This is due to `noImplicitReturns`.
I'm OK with disabling it, if requested.

But I think the code is actually more readable and you will never forget to return something or miss that its not returning when reviewing.

Also, I prefer the current code, it might have more lines but its more readable.

[the remark this answers is not in the document]
Comment (Devp00l) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: We should vote for it during the next stand up :)
I'm for disabling it, as how a method/function works should be known.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/pwd-expiration-notification/pwd-expiration-notification.component.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (Devp00l) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/components/pwd-expiration-notification/pwd-expiration-notification.component.ts]: If `pwdExpirationData` is not set you will call `Date` with `NaN` this would return 'NaN' at the end. I'm not sure if it will break something but it has much more potential to do so.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (LenzGr) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: > But I think the code is actually more readable and you will never forget to return something or miss that its not returning when reviewing.

FWIW, I too am in favor of better readability. Keep in mind that we should write code that others could quickly get familiar with in order to lower the barrier for new contributors (and for reviewers, FWIW).

I suggest to keep the check for implicit returns - this also improves the readability and avoids ambiguity.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-configuration-form/rbd-configuration-form.component.ts [hunk 15]
```diff
@@ -1,6 +1,8 @@
 import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
 import { FormControl, Validators } from '@angular/forms';
 
+import * as _ from 'lodash';
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-configuration-form/rbd-configuration-form.component.ts]: You forgot to remove the import.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (p-se) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: I tend to be more strict as well. I like the `noImplicitReturns` rule and I think (and hope) it will eventually lead to code that looks like it was written in a strictly typed language. That is, when declared to return a number, will return a number. Taking the example, it'd look like this:

```typescript
private replicatedPgCalc(pgs: number): number {
  const sizeControl = this.form.get('size');
  const size = sizeControl.value;
  if (sizeControl.valid && size > 0) {
    return pgs / size;
  }
  return 0; // or: -1   <---
}
```

As alternative, the method/function can throw an exception which can be caught by the caller.

This looks way more natural to me than writing `return undefined`, which by the way, feels unnatural to me for two reasons:

1. The method declared a number to be returned but returned nothing
2. `undefined` is the return value for attributes of objects which aren't there or method/function arguments that haven't been provided. I'd prefer to use `null` if necessary, but would even try to avoid that by either returning a variable of the declared type or raising an exception.

Either way, **the way to improve in that regard is to keep `noImplicitReturns`** and eventually write code that doesn't feel so unnatural. But as a first step I wouldn't require the code to be changed this way in this PR, as this is the current behavior the code and the PR is really big enough, even without changing its behavior.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/pwd-expiration-notification/pwd-expiration-notification.component.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (Devp00l) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/components/pwd-expiration-notification/pwd-expiration-notification.component.ts]: I tested it locally doesn't seem to break anything.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (Devp00l) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: Any changes can me made in a follow up PR to get it in quickly.
